nixos/modules/services/security/vault-agent.nix

```diff
@@ -0,0 +1,143 @@
+{config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  cfg = config.services.vault-agent;
+  opt = options.services.vault-agent;
+  createAgentInstance = cfg: name:
+    let
+      configFile = pkgs.writeText "${name}.hcl" ''
+          pid_file = "/run/${name}.pid"
+          vault {
+            address = "${cfg.address}"
+          }
+          ${cfg.extraConfig}
+          ${mkTemplates cfg.templates}
+        '';
+      mkTemplates = templates: concatStringsSep "\n" (map (config: ''
+      template {
+        source      = "${mkTemplateFiles config}"
+        command     = "${config.command}"
+        destination = "${config.destination}"
+        perms       = "${config.perms}"
+      }
+      '') (mapAttrsToList (k: v: v // { template = k; }) templates));
+
+      mkTemplateFiles = config: pkgs.writeText "${config.template}" ''${config.content}'';
+    in  {
+      description = "Vault-agent ${name} daemon";
+      wantedBy = ["multi-user.target"];
+      after = [ "network.target" ];
+      restartIfChanged = true;
+      startLimitIntervalSec = 60;
+      startLimitBurst = 3;
+      serviceConfig = {
+        User = "root";
+        Group = "root";
```

Comment:
In fact I'll remove all the user part like you said because this service would indeed require root user. 
The point of it is to put some of your vault data in specific location and interacts with systemd services so yes root is required.